src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php

```diff
@@ -87,4 +83,16 @@ public function process(ContainerBuilder $container)
             }
         }
     }
+
+    private function validateTagAttributes(string $id, string $tag, array $attributes, array $path = []): void
+    {
+        foreach ($attributes as $name => $value) {
+            if (\is_array($value)) {
+                $this->validateTagAttributes($id, $tag, $value, [...$path, $name]);
+            } elseif (!\is_scalar($value) && null !== $value) {
+                $name = implode('.', [...$path, $name]);
+                throw new RuntimeException(sprintf('A "tags" attribute must be of a scalar-type for service "%s", tag "%s", attribute "%s".', $id, $tag, $name));
```

Comment:
I chose to report  the path to the offending value here because that's what was done in #47364.
I find this confusing however. For example in that case:
```php
$services->set('example', Example::class)
  ->tag('a', ['foo.bar.baz' => new class{}])
  ->tag('a', ['foo' => ['bar' => ['baz' => new class{}]]);
```
The error message would be:
```
A "tags" attribute must be of a scalar-type for service "example", tag "a", attribute "foo.bar.baz".
```

IMO it would be better to just report the top-level attribute name.

WDYT?

Reply:
> Code wise, let's use a string instead of an array for the path

Well, the code in the aforementioned PR is actually incorrect...
For the previous example, It reports 'bar.baz' instead of 'foo.bar.baz'.
IMO using a stack is clearer and simpler (less conditionals), but I won't die for this idea.

Reply:
Can you fix the code in this PR also please? not sure what you mean about the complexity of using string, I see none :)
Can you give it a try?

Reply:
The alternative would look like this:
```php
$this->validateTagAttributes($id, $tag, $value, '' === $path ? $name : "{$path}.{$name}");
```
Which is longer, less declarative and less readable to me.
Do you confirm you want this?

Reply:
> not sure what you mean about the complexity of using string

Well, implementing this using a string actually took me two attempts to get it right (vs just one using a stack).
And I believe it to be the reason why #47364 got it wrong in the first place.

> Can you fix the code in this PR also please?

Sure can.  Before I can proceed, I just need a confirmation on this style-related point.

Reply:
> Can you fix the code in this PR also please?

Also, if I'm going to fix this issue, I might as well just report the top-level attribute inspeat of the full path to the value.

Reply:
got it, let's go with the stack everywhere!